thavalon-server/src/connections/validation.rs

```diff
@@ -0,0 +1,257 @@
+use chrono::{Duration, Utc};
+use jsonwebtoken::{DecodingKey, EncodingKey, Header, Validation};
+use rand::{distributions::Alphanumeric, Rng};
+use scrypt::{errors::CheckError, ScryptParams};
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::env;
+use std::iter;
+use std::sync::{Arc, Mutex};
+use thiserror::Error;
+
+const PASSWORD_MIN_LENGTH: usize = 8;
+
+#[derive(Debug, Error, Eq, PartialEq)]
+pub enum ValidationError {
+    #[error("Fatal hashing error")]
+    HashError,
+    #[error("Invalid password")]
+    InvalidPassword,
+    #[error("User unathorized for this request.")]
+    Unauthorized,
+}
+
+#[derive(Serialize)]
+pub struct JWTResponse {
+    token_type: String,
+    access_token: String,
+    expires_at: i64,
+}
+
+#[derive(Clone)]
+pub struct RefreshTokenInfo {
+    pub token: String,
+    pub expires_at: i64,
+    pub email: String,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct JWTClaims {
+    aud: String,
+    exp: i64,
+    iat: i64,
+    iss: String,
+    nbf: i64,
+    sub: String,
+}
+
+pub type TokenStore = Arc<Mutex<HashMap<String, RefreshTokenInfo>>>;
+
+/// Hashes a plaintext password using the currently selected hashing algorithm.
+///
+/// # Arguments
+///
+/// * `plaintext` - the plain text password to be hashed
+///
+/// # Returns
+///
+/// * `Result<password_hash, error>`
+pub async fn hash_password(plaintext: &String) -> Result<String, ValidationError> {
+    if plaintext.len() < PASSWORD_MIN_LENGTH {
+        log::warn!("Received a password below minimum security specs");
+        return Err(ValidationError::InvalidPassword);
+    }
+
+    let hash = scrypt::scrypt_simple(plaintext, &ScryptParams::recommended()).map_err(|e| {
+        log::error!("An RNG error occurred with the underlying OS.");
+        log::error!("{:?}", e);
+        ValidationError::HashError
+    });
+
+    hash
+}
+
+/// Validates a plaintext password against a given hash.
+///
+/// # Arguments
+///
+/// * `plaintext` - the plain text password to check
+/// * `hash` - Password hash in scrypt format
+///
+/// # Returns
+/// True if passwords match. False otherwise.
+pub async fn validate_password(plaintext: &String, hash: &String) -> bool {
+    let result = match scrypt::scrypt_check(plaintext, hash) {
+        Ok(_) => true,
+        Err(e) => {
+            if e == CheckError::InvalidFormat {
+                log::error!("Database hash is not in a valid scrypt format.");
+            }
+            false
+        }
+    };
+
+    result
+}
+
+/// Creates a valid JWT using a user's email.
+///
+/// # Arguments
+///
+/// * `user_email` - The user's email to create a JWT with.
+///
+/// # Returns
+///
+/// A JWTResponse with the JWT
+pub async fn create_jwt(
+    user_email: &String,
+    token_store: TokenStore,
+) -> (JWTResponse, RefreshTokenInfo) {
+    log::info!("Creating a new JWT for {}.", user_email);
+    let jwt_secret = env::var("JWT_SECRET").unwrap_or("JWT_SECRET".to_string());
```

Comment:
Having a custom TokenVar struct would also let us cache this lookup, or even the parsed keys

Reply:
I forgot I had these lines. Meant to change them to a static but forgot. I've moved them to a static variable. I think that the key is going to change infrequently enough that we can use a static and change when the machine shuts down and restarts if needed.